### PR TITLE
Python 3.4/5/6 and PyPy support. 

### PR DESCRIPTION
**Pull request by jamadden**
**Needs a release of zope.app.renderer and zope.app.preference**

- Add modern tox.ini and .travis.yml, with coveralls support.
- 99% test coverage.
- Rename .txt -> .rst

### CODE REVIEW

File: src/zope/app/onlinehelp/onlinehelptopic.py
Comment (jamadden) [inline src/zope/app/onlinehelp/onlinehelptopic.py:164]: Hmm, this can apparently cause issues with security checkers in Python 3.5+, after OrderedDict was re-implemented in C:
```
  File "//lib/python3.5/site-packages/zope/app/tree/node.py", line 95, in getChildObjects
    children = self._get_child_objects_adapter().getChildObjects()
  File "/lib/python3.5/site-packages/zope/app/tree/adapters.py", line 94, in getChildObjects
    return list(self.context.values()) if self.hasChildren() else []
zope.security.interfaces.ForbiddenAttribute: ('__iter__', odict_values([<zope.app.onlinehelp.onlinehelptopic.RESTOnlineHelpTopic object at 0x10f87aac8>, <zope.app.onlinehelp.onlinehelptopic.RESTOnlineHelpTopic object at 0x10f87a128>]))
```

Comment (jamadden) [reply src/zope/app/onlinehelp/onlinehelptopic.py:164]: I suspect that's probably more of a problem in zope.security's default allow lists, though.

Comment (jamadden) [reply src/zope/app/onlinehelp/onlinehelptopic.py:164]: FWIW, this code fixes it and should probably be added to zope.security:
```python
from zope.security import checker
checker._default_checkers[type(OrderedDict().values())] = checker._iteratorChecker
checker._default_checkers[type(iter(OrderedDict().values()))] = checker._iteratorChecker
```

Comment (icemac) [reply src/zope/app/onlinehelp/onlinehelptopic.py:164]: @jamadden Do you mind to push this code forward to `zope.security`?

**Comment by mgedmin** [comment]
You now have PyPI rights.

File: src/zope/app/onlinehelp/browser/tests.py
Comment (mgedmin) [inline src/zope/app/onlinehelp/browser/tests.py]: A very good question!  I don't know/remember.

I wouldn't be surprised if the answer is some magic ++help++ traversal scheme that uses LocationProxies to make context/@@absolute_url work.  (It's a blind guess.)

Comment (mgedmin) [reply src/zope/app/onlinehelp/browser/tests.py]: Oh God, it's worse than that.

There's a ++help++ (`helpNamespace`), but it doesn't use location proxies.  It just assigns to the `context` attribute of a shared global singleton OnlineHelp instance!  Threads?  Never heard'o'em.

Comment (mgedmin) [reply src/zope/app/onlinehelp/browser/tests.py]: Not to mention having module globals keep references to persistent objects loaded using a ZODB connection from an arbitrary thread, kept alive after the transaction that loaded them is closed, is just Not A Good Idea.

**Review by mgedmin — APPROVED** [review]
LGTM, once the XXXes are resolved.

I don't know if the XXX in the test needs to be resolved.  Some judicious use of LocationProxy() in the test might suffice maybe?

Fixing the helpNamespace shenanigans is IMHO out of scope for this PR.

File: src/zope/app/onlinehelp/onlinehelptopic.py
Comment (jamadden) [reply src/zope/app/onlinehelp/onlinehelptopic.py:164]: Yes, I will open a PR.

Comment (jamadden) [reply src/zope/app/onlinehelp/onlinehelptopic.py:164]: https://github.com/zopefoundation/zope.security/pull/24

File: src/zope/app/onlinehelp/browser/tests.py
Comment (jamadden) [reply src/zope/app/onlinehelp/browser/tests.py]: > There's a ++help++ (helpNamespace), but it doesn't use location proxies. It just assigns to the context attribute of a shared global singleton OnlineHelp instance

Ick. But maybe easy enough to emulate locally in testing ~~since we have broken the (cyclic) dependency on zope.app.apidoc~~. I'll take a look.

> Threads? Never heard'o'em.

~~It's only enabled when `devmode` is, right? I wouldn't be surprised if that also tends to single-thread requests.~~

EDIT: Yeah, not sure why I was conflating `++apidoc++` and `++help++` there.

Comment (jamadden) [reply src/zope/app/onlinehelp/browser/tests.py]: > It just assigns to the context attribute of a shared global singleton OnlineHelp instance!

That is weird, but I think this was mainly a misunderstanding on my part.

The TAL expression that's failing is looking for `topicContent` on the *view*: `tal:content="structure view/topicContent"`. This works for the *other* `TopicView` class `OnlineHelpTopicView` because it defines a `topicContent` method, but `ZPTOnlineHelpTopicView` does not.

I think the underlying issue here is that `ZPTOnlineHelpTopicView` isn't really meant to be used with the `helptopic.pt` template file. You need to create traversable instances of `ZPTOnlineHelpTopic` with a `path` to a  template file, and when you traverse to these they get rendered. 

(I wonder if this is supposed to be done TTW [which seems weird, because you have to have a file on disk], but that's not demonstrated.)

**Comment by jamadden** [comment]
The blocking dependencies have been released, and the XXX resolved. I would like to wait a bit before merging and releasing, though, in case zopefoundation/zope.security#23 gets resolved and we can declare a dependency on the fix here (since this is where the issue is introduced).

**Comment by jamadden** [comment]
Thank you!

File: src/zope/app/onlinehelp/browser/tests.py
Comment (mgedmin) [reply src/zope/app/onlinehelp/browser/tests.py]: I'm quite sure no TTW is involved; help pages are all registered via ZCML directives.

(I'm not familiar with the implementation details such as what the `ZPTOnlineHelpTopicView` class is supposed to do, or what's `helptopic.pt` was meant for.)

**Pull request merged**
